crates/bevy_ecs/src/query/point_query.rs

```diff
@@ -0,0 +1,186 @@
+    component::Tick,
+    entity::{Entities, Entity},
+    storage::{TableId, Tables},
+    world::unsafe_world_cell::UnsafeWorldCell,
+};
+
+use super::{DebugCheckedUnwrap, QueryData, QueryEntityError, QueryFilter, QueryState, StorageId};
+
+/// Used for quickly fetching query item from a specific [`Entity`].
+/// It caches the last fetch which could potentially be more efficient when dealing with many entities of the same archetype.
+///
+/// This struct is created by the [`Query::as_point_query`](crate::system::Query::as_point_query) and [`Query::as_point_query_mut`](crate::system::Query::as_point_query_mut) methods.
+pub struct PointQuery<'w, 's, D: QueryData, F: QueryFilter = ()> {
+    // SAFETY: Must have access to the components registered in `state`.
+    entities: &'w Entities,
+    tables: &'w Tables,
+    archetypes: &'w Archetypes,
+    state: &'s QueryState<D, F>,
+    fetch: D::Fetch<'w>,
+    filter: F::Fetch<'w>,
+    last_storage_id: StorageId,
+}
+
+impl<'w, 's, D: QueryData, F: QueryFilter> PointQuery<'w, 's, D, F> {
+    const IS_DENSE: bool = D::IS_DENSE && F::IS_DENSE;
+    /// # Safety
+    /// - `world` must have permission to access any of the components registered in `query_state`.
+    /// - `world` must be the same one used to initialize `query_state`.
+    #[inline]
+    pub(crate) unsafe fn new(
+        world: UnsafeWorldCell<'w>,
+        query_state: &'s QueryState<D, F>,
+        last_run: Tick,
+        this_run: Tick,
+    ) -> PointQuery<'w, 's, D, F> {
+        let fetch = D::init_fetch(world, &query_state.fetch_state, last_run, this_run);
+        let filter = F::init_fetch(world, &query_state.filter_state, last_run, this_run);
+        let last_storage_id = if Self::IS_DENSE {
+            StorageId {
+                table_id: TableId::INVALID,
+            }
+        } else {
+            StorageId {
+                archetype_id: ArchetypeId::INVALID,
+            }
+        };
+
+        PointQuery {
+            state: query_state,
+            entities: world.entities(),
+            archetypes: world.archetypes(),
+            // SAFETY: We only access table data that has been registered in `query_state`.
+            // This means `world` has permission to access the data we use.
+            tables: &world.storages().tables,
+            fetch,
+            filter,
+            last_storage_id,
+        }
+    }
+
+    /// Returns the read-only query item for the given [`Entity`].
+    ///
+    /// In case of a nonexisting entity or mismatched component, a [`QueryEntityError`] is returned instead.
+    ///
+    /// This is always guaranteed to run in `O(1)` time.
+    #[inline]
+    pub fn get(&mut self, entity: Entity) -> Result<D::Item<'_>, QueryEntityError> {
+        // SAFETY: system runs without conflicts with other systems.
+        // same-system queries have runtime borrow checks when they conflict
+        unsafe { std::mem::transmute(self.get_unsafe(entity)) }
```

Comment:
```suggestion
        let result = unsafe { self.get_unsafe(entity) };
        match result {
            Ok(item) => Ok(D::shrink(item)),
            Err(err) => Err(err),
        }
```

We can use [`WorldQuery::shrink`](https://docs.rs/bevy/0.14.1/bevy/ecs/query/trait.WorldQuery.html#tymethod.shrink) instead of a transmute.